### ISSUE TITLE
Trying get the digest for empty byte array produces scary errors

### ISSUE THREAD

**Issue by olizilla** [web3-storage/fr32-sha2-256-trunc254-padded-binary-tree-multihash#13]
e.g 
```
    unreachable

    at data:application/wasm;base64,AGFzbQEAAAABaRBgAn9/AX9gAn9/AGADf39/AX9gA39/fwBgAX8AYAR/f39/AGAFf39/f38AYAV/f39/fwF/YAAAYAR/f39/AX9gAAF/YAF/AX9gB39/f39/f38Bf2ACfn8Bf2AHf39/f39/fwBgAX8BfgJeAwN3YmceX193YmluZGdlbl9jb3B5X3RvX3R5cGVkX2FycmF5AAMDd2JnGl9fd2JpbmRnZW5fb2JqZWN0X2Ryb3BfcmVmAAQDd2JnEF9fd2JpbmRnZW5fdGhyb3cAAQNsawMCAAEDBwIHAQMMAA0AAAYFAQUFAQADBgMFBQEADgEFAQYFAwMBBgAHAQEDAwMDBAEEAQMEAgIBAQEADwcIAQkBAAQAAAoEBgEEBAAAAAYBAAkBAQEAAQAAAAAEAAABCAgBAAMCCgALAAsEBAUBcAExMQUDAQARBgkBfwFBgIDAAAsHqgEJBm1lbW9yeQIAFl9fd2JnX3BpZWNlaGFzaGVyX2ZyZWUANBFwaWVjZWhhc2hlcl9jb3VudAA+EXBpZWNlaGFzaGVyX3Jlc2V0AEwRcGllY2VoYXNoZXJfd3JpdGUAJxZwaWVjZWhhc2hlcl9kaWdlc3RJbnRvACAGY3JlYXRlAEgScGllY2VoYXNoZXJfY3JlYXRlAEgRX193YmluZGdlbl9tYWxsb2MAPQk2AQBBAQswbS1LX1JdXlNcRURtImlqbSVrbDNtOA5GTVVSOldJFCFtVm05WkdtTltQWE8qEW1WCvuAAmv2NQEqfyMAQZABayIDJAAgA0HAABBlIQMgACgCHCEoIAAoAhghKSAAKAIUISQgACgCECElIAAoAgwhKiAAKAIIISsgACgCBCEmIAAoAgAhJyACBEAgASACQQZ0aiEsA0BBACECA0AgAiADaiABIAJqKAAAIhRBGHQgFEGA/gNxQQh0ciAUQQh2QYD+A3EgFEEYdnJyNgIAIAJBBGoiAkHAAEcNAAsgAygCPCECIAMoAjghFCADKAI0IRUgAygCMCEWIAMoAiwhGyADKAIoIRwgAygCJCEdIAMoAiAhHiADKAIcIR8gAygCGCEgIAMoAhQhISADKAIQISIgAygCDCEXIAMoAgghGCADKAIEIRkgAygCACEaIAMgKzYCYCADICo2AmQgAyApNgJoIAMgKDYCbCADICQ2AnwgAyAlNgJ4IAMgJjYCdCADICc2AnAgAyAaQZjfqJQEaiIENgKMASADIBlBkYndiQdqIgU2AogBIAMgGEGxiPzRBGsiBjYChAEgAyAXQdvIqLIBayIHNgKAASADQdAAaiIMIANB4ABqIg0gA0HwAGoiDiADQYABaiIPEBMgAygCUCEIIAMoAlQhCSADKAJYIQogAygCXCELIAMgJDYCbCADICU2AmggAyAmNgJkIAMgJzYCYCADIAs2AnwgAyAKNgJ4IAMgCTYCdCADIAg2AnAgAyAGNgKMASADIAc2AogBIAMgBDYChAEgAyAFNgKAASAMIA0gDiAPEBMgAygCUCEEIAMoAlQhBSADKAJYIQYgAygCXCEHIAMgCzYCbCADIAo2AmggAyAJNgJkIAMgCDYCYCADIAc2AnwgAyAGNgJ4IAMgBTYCdCADIAQ2AnAgAyAiQduE28oDaiIQNgKMASADICFB8aPEzwVqIhE2AogBIAMgIEHc+oHuBmsiEjYChAEgAyAfQavCjqcFayITNgKAASAMIA0gDiAPEBMgAygCUCEIIAMoAlQhCSADKAJYIQogAygCXCELIAMgBzYCbCADIAY2AmggAyAFNgJkIAMgBDYCYCADIAs2AnwgAyAKNgJ4IAMgCTYCdCADIAg2AnAgAyASNgKMASADIBM2AogBIAMgEDYChAEgAyARNgKAASAMIA0gDiAPEBMgAygCUCEEIAMoAlQhBSADKAJYIQYgAygCXCEHIAMgCzYCbCADIAo2AmggAyAJNgJkIAMgCDYCYCADIAc2AnwgAyAGNgJ4IAMgBTYCdCADIAQ2AnAgAyAeQeiq4b8CayIQNgKMASADIB1BgbaNlAFqIhE2AogBIAMgHEG+i8ahAmoiEjYChAEgAyAbQcP7sagFaiITNgKAASAMIA0gDiAPEBMgAygCUCEIIAMoAlQhCSADKAJYIQogAygCXCELIAMgBzYCbCADIAY2AmggAyAFNgJkIAMgBDYCYCADIAs2AnwgAyAKNgJ4IAMgCTYCdCADIAg2AnAgAyASNgKMASADIBM2AogBIAMgEDYChAEgAyARNgKAASAMIA0gDiAPEBMgAygCUCEEIAMoAlQhBSADKAJYIQYgAygCXCEHIAMgCzYCbCADIAo2AmggAyAJNgJkIAMgCDYCYCADIAc2AnwgAyAGNgJ4IAMgBTYCdCADIAQ2AnAgAyAWQfS6+ZUHaiIQNgKMASADIBVBgpyF+QdrIhE2AogBIAMgFEHZ8o+hBmsiEjYChAEgAyACQYydkPMDayITNgKAASAMIA0gDiAPEBMgAygCUCEIIAMoAlQhCSADKAJYIQogAygCXCELIAMgBzYCbCADIAY2AmggAyAFNgJkIAMgBDYCYCADIAs2AnwgAyAKNgJ4IAMgCTYCdCADIAg2AnAgAyASNgKMASADIBM2AogBIAMgEDYChAEgAyARNgKAASAMIA0gDiAPEBMgAygCUCEEIAMoAlQhBSADKAJYIQYgAygCXCEHIAMgGjYCXCADIBk2AlggAyAYNgJUIAMgFzYCUCADICI2AmwgAyAhNgJoIAMgIDYCZCADIB82AmAgAyAeNgJ8IAMgHTYCeCADIBw2AnQgAyAbNgJwIAMgFjYCjAEgAyAVNgKIASADIBQ2AoQBIAMgAjYCgAEgA0FAayIjIAwgDSAOIA8QEiADKAJAIRcgAygCRCEYIAMoAkghGSADKAJMIRogAyALNgJsIAMgCjYCaCADIAk2AmQgAyAINgJgIAMgBzYCfCADIA...
```

I hit this when i accidentally tried to get the digest for a "zero byte car file" (a mistake) but i think it should probably throw an less scary error.

**Comment by olizilla** [comment]
> The [crate console_error_panic_hook](https://crates.io/crates/console_error_panic_hook) allows debugging Rust panic messages in a web browser, making it much easier to debug WebAssembly code.
>
> Let's compare what happens when Rust code panics before and after enabling the feature:
>
> Before: "RuntimeError: Unreachable executed"
>
> After: "panicked at 'index out of bounds: the len is 3 but the index is 4', libcore/slice/mod.rs:2046:10"
>
> To do this, a [panic hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html) is configured that logs panics to the developer console via the JavaScript console.error function.
>
> Note though that console_error_panic_hook is not entirely automatic, so you'll need to make sure that utils::set_panic_hook() is called before any of our code runs (and it's safe to run set_panic_hook many times).
>
> For more details, see the [console_error_panic_hook repository](https://github.com/rustwasm/console_error_panic_hook).

– https://rustwasm.github.io/docs/wasm-pack/tutorials/npm-browser-packages/template-deep-dive/src-utils-rs.html#2-what-is-console_error_panic_hook

**Comment by olizilla** [comment]
I suspect the solution here will be to change the error handling for "too few bytes" from panic to returning an error result. 
https://doc.rust-lang.org/book/ch09-00-error-handling.html

**Comment by Gozala** [comment]
I suspect you’re hitting this

https://github.com/web3-storage/fr32-sha2-256-trunc254-padded-binary-tree-multihash/blob/3acb44a333b6ca52a1c62aa8ce3fef824423117e/src/hasher.rs#L113-L117

And you are right idiomatic rust won’t panic but rather utilize a result type. Which I have not used for api compatibility with multihashhasher interface expected by most js code.

Fundamental issue here is that this hasher imposes min payload size, which no other hasher does. I raised this problem when reviewing the spec but it still been discussed.

Solution here is probably finding a way how to make rust throw specific errors on panic as opposed to just hitting unreachable. I suspect wasm-bindgen has ways to do it but I have not looked into as I assumed it was the default behavior

**Comment by Gozala** [comment]
@olizilla I've looked into if there was to throw error from rust-wasm, and turns out wasm_bindgen does that by default if your function returns Result type see https://rustwasm.github.io/wasm-bindgen/reference/types/result.html

That is to say if we just switch functions that panic to return Result's it will have desired behavior in JS.

**Comment by Gozala** [comment]
Was fixed in 2.0 by lifting min size limit per https://github.com/filecoin-project/FIPs/pull/808

There is a test case for it now too

https://github.com/web3-storage/fr32-sha2-256-trunc254-padded-binary-tree-multihash/blob/f88a0dc88c89fc96133086ad24b01f4ae756f7d1/test/lib.spec.js#L67-L121
